### PR TITLE
DAP-40230: Changes to be able to build a spark image to run datameer …

### PR DESCRIPTION
**Pull request by mbauhardt**
…jobs against GKE cluster

### CODE REVIEW

File: dev/export-datameer-env.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export __vHadoop='hadoop-2.7'
```
Comment (mfuhrmeisterDM) [inline dev/export-datameer-env.sh:1]: why 2.7? Isn't spark-2.4.2 based on hadoop-2.6.5?

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile [hunk 2]
```diff
@@ -51,4 +51,11 @@ ENV SPARK_HOME /opt/spark
 
 WORKDIR /opt/spark/work-dir
 
+# add the following two lines before ENTRYPOINT
+RUN rm $SPARK_HOME/jars/kubernetes-client-*.jar
+RUN rm $SPARK_HOME/jars/guava-14.0.1.jar
+ADD https://repo1.maven.org/maven2/io/fabric8/kubernetes-client/4.4.2/kubernetes-client-4.4.2.jar $SPARK_HOME/jars
```
Comment (mfuhrmeisterDM) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile:57]: 4.4.2? why this version and not the one we have in dm?

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile [hunk 3]
```diff
@@ -51,4 +51,11 @@ ENV SPARK_HOME /opt/spark
...
+ADD https://repo1.maven.org/maven2/com/google/guava/guava/23.0/guava-23.0.jar $SPARK_HOME/jars
```
Comment (mfuhrmeisterDM) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile:58]: @kkolman

**Review by mfuhrmeisterDM — COMMENTED** [review]
For me a proper dependency resolution is missing. We see this problem currently in normal spark project. See my last changeset or fabians or karels current ticket. I would like to have a generated jars dir depending on maven dependencies. 
@mbauhardt @kkolman @peri4n can we have a call about this tomorrow?

File: dev/export-datameer-env.sh (hunk 1 above)
Comment (mbauhardt) [reply dev/export-datameer-env.sh:1]: you can choose between them, we chosed hadoop2.7 for the image we worked with

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile (hunk 3 above)
Comment (peri4n) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile:58]: Spark 2.4.2 needs a Guava that is <= 16.0 .

File: dev/export-datameer-env.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+export __vScala='scala-2.12'
+export __vDatameer='datameer-11.1-gcs'
+export __distSuffix=${__vHadoop}-${__vScala}-${__vDatameer}
+export __dockerTag='2.4.2'-${__distSuffix}
+export __dockerRegistry=us.gcr.io/devops-project-200915
```
Comment (MikeSchulze) [inline dev/export-datameer-env.sh:6]: if the names given by spark env?
It looks a bit unusual for environment variables

**Review by MikeSchulze — APPROVED** [review]

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile (hunk 2 above)
Comment (mbauhardt) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile:57]: yes. I had exactly the same question. but was somehow to lazy to get it why. i copied over the lines from the image definition we used until today.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile (hunk 3 above)
Comment (mbauhardt) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile:58]: same answer I gave above.

File: dev/export-datameer-env.sh (hunk 1 above)
Comment (peri4n) [reply dev/export-datameer-env.sh:1]: It is.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile [hunk 5]
```diff
@@ -51,4 +51,11 @@ ENV SPARK_HOME /opt/spark
...
+RUN cp /lib/ld-musl-x86_64.so.1 /lib/ld-linux-x86-64.so.2
```
Comment (peri4n) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile:59]: Why is this necessary?

File: dev/export-datameer-env.sh (hunk 4 above)
Comment (mbauhardt) [reply dev/export-datameer-env.sh:6]: no just used us. i choosed the double underscore to make sure noone else is using the same vars.

**Comment by peri4n** [comment]
@mfuhrmeisterDM agreed.
@mbauhardt Please don't let us go back und commit jar files into git again.

**Review by peri4n — CHANGES_REQUESTED** [review]

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile (hunk 3 above)
Comment (kkolman) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile:58]: @peri4n wasnt the guava version (and the conflicts) requirement of the hadoop-2.6.5 deps ?

because as far as i can see
guava is a provided dependency, so you "should be able to swap in whichever you like"
https://mvnrepository.com/artifact/org.apache.spark/spark-core_2.12/2.4.4

and the whichever you like would be the one coming from the "hadoop" version

but of course dependency hell is purely theoriticaly right until it fails in production...

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile (hunk 5 above)
Comment (mbauhardt) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile:59]: @mfuhrmeisterDM  and me ran into the situation that the linux so files can't be found. the solution was to link or copy the musl libc implementation into a file which can be found.
but I can cross check again.

File: dev/export-datameer-env.sh (hunk 1 above)
Comment (mbauhardt) [reply dev/export-datameer-env.sh:1]: https://archive.apache.org/dist/spark/spark-2.4.2/
you can choose both

**Comment by mbauhardt** [comment]
we don't wanna go this way
